core/src/main/java/org/elasticsearch/index/query/FilteredQueryBuilder.java

```diff
@@ -40,6 +46,15 @@
     static final FilteredQueryBuilder PROTOTYPE = new FilteredQueryBuilder(null, null);
 
     /**
+     * A query that applies a filter to the results of a match_all query.
+     * @param filterBuilder The filter to apply on the query (Can be null)
+     * */
+    public FilteredQueryBuilder(@Nullable QueryBuilder filterBuilder) {
+        this.queryBuilder = new MatchAllQueryBuilder();
+        this.filterBuilder = filterBuilder;
+    }
+
```

Comment:
I think the question here boils down to how much of the optimization/default handling that was previously done in the parser should be moved to the builder and thus made available to users of the Java API.

If we want Java API and REST API to be similar so users can directly map from what they are doing with e.g. Sense to Java API calls I believe we should be lenient wrt. accepting null values in the Builder.

If we'd rather avoid handling null values in the Builder at all the resulting Java API will be more strict than what we accept through REST.

In the latest update to the PR I tried to find a middle ground - accepting null values but mapping them to EmptyQueryBuilder where possible.